src/v/archival/archiver_operations_impl.cc

```diff
@@ -41,6 +41,33 @@ namespace archival {
 namespace detail {
 
 class archiver_operations_impl : public archiver_operations_api {
+    /// Combined result of several uploads
+    struct aggregated_upload_result {
+        cloud_storage::upload_result code{
+          cloud_storage::upload_result::success};
+        /// Populated on success
+        std::optional<cloud_storage::segment_record_stats> stats;
+        /// Total number of PUT requests used
+        size_t put_requests{0};
+        /// Total number of bytes sent
+        size_t bytes_sent{0};
+
+        // Combine aggregated results
+        ///
+        /// The error codes are combined using max function (the worst
+        /// error wins). The order is: success < timeout < failure < cancelled.
+        /// The 'stats' objects can't be combined so the series of aggregated
+        /// results can't contains more than one 'stats' instances.
+        void combine(const aggregated_upload_result& other) {
+            put_requests += other.put_requests;
+            bytes_sent += other.bytes_sent;
+            if (!stats.has_value()) {
+                stats = other.stats;
+            }
+            code = std::max(code, other.code);
```

Comment:
This seems like it will be error prone. Should there be a container of bad codes? Or maybe we should have a  container of some segment result type

```
struct segment_upload_result{
  upload_result code;
  std::optional<segment_record_stats> stats;
  ...
};
```

Reply:
Why is it error prone? The only reason why this could lead to error is if someone will change success error code (currently it's zero).